### PR TITLE
Added PBR Shader

### PR DESCRIPTION
**Pull request by bonjorno7**
there's some small bugs but I dunno how to fix them so this will do for now

### CODE REVIEW

**Comment by SlidyBat** [comment]
Do you mind listing the remaining bugs in detail, just so it’s “documented” somewhere.

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 1]
```diff
@@ -0,0 +1,524 @@
...
+        bool bHasEmissive       = (info.Emissive != -1) && params[info.Emissive]->IsTexture();
+        bool bHasEnvmap         = (info.Envmap != -1) && params[info.Envmap]->IsTexture();
+        bool bHasFlashlight     = UsingFlashlight(params);
+        bool bLightmapped       = IS_FLAG_SET(MATERIAL_VAR_MODEL) == 0;
+
+        // determining whether we're dealing with a fully opaque material
+        BlendType_t nBlendType = this->EvaluateBlendRequirements(info.Albedo, true);
+        bool bFullyOpaque = (nBlendType != BT_BLENDADD) && (nBlendType != BT_BLEND) && !bIsAlphaTested;
+
+        if (this->IsSnapshotting()) // shadow state
```
Comment (SlidyBat) [inline mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: I'd prefer if you used the `SHADOW_STATE`/`DYNAMIC_STATE` macros

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+            int nShadowFilterMode = bHasFlashlight ? g_pHardwareConfig->GetShadowFilterMode() : 0;
+
+            // if the albedo texture exists, set the default blendding shadow state
+            if (params[info.Albedo]->IsTexture())
+            {
+                SetDefaultBlendingShadowState(info.Albedo, true);
+            }
+
+            // setting up samplers
+            pShaderShadow->EnableTexture(SHADER_SAMPLER0, true);    // Albedo Texture
```
Comment (SlidyBat) [inline mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: Perhaps make some defines up top that map names to sampler slots? E.g. `#define ALBEDO_SAMPLER SHADER_SAMPLER0`.

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+
+            // enable alpha writes all the time so that we have them for underwater stuff
+            pShaderShadow->EnableAlphaWrites(bFullyOpaque);
+        }
+        else // not snapshotting - begin dynamic state
+        {
+            // setting up albedo texture
+            if (bHasAlbedo)
+            {
+                this->BindTexture(SHADER_SAMPLER0, info.Albedo, info.BaseTextureFrame);
```
Comment (SlidyBat) [inline mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: Drop the `this->`, it's inconsistently used across the file, and IMO isn't needed.

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+
+// Get lightmap color
+float3 lightmapLookup(float3 textureNormal, float4 lightmapTexCoord1And2, float4 lightmapTexCoord3)
+{
+	float2 bumpCoord1;
+	float2 bumpCoord2;
+	float2 bumpCoord3;
+
+	ComputeBumpedLightmapCoordinates(
+			lightmapTexCoord1And2, lightmapTexCoord3.xy,
```
Comment (SlidyBat) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Lots of space/tab mixing in this file, please change it to be consistent

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+			fresnelReflectance, roughness, metalness, lightDirectionAngle, albedo, true);
+#else
+        directLighting += calculateLight(LightIn, LightColor, outgoingLightDirection, normal,
+			fresnelReflectance, roughness, metalness, lightDirectionAngle, albedo, false);
+#endif
+    }
+	// end direct
+
+	// start flashlight
+	if (FLASHLIGHT > 0)
```
Comment (SlidyBat) [inline mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: Better off as a `#if`

File: mp/src/materialsystem/stdshaders/pbr_vs20.fxc [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+// includes
+#include "common_vs_fxc.h"
+
+// shader specific constants
+const float4 cBaseTexCoordTransform[2] : register(SHADER_SPECIFIC_CONST_0);
+
+// shader input
+struct VS_INPUT
+{
+	// this is all of the stuff that we ever use
```
Comment (SlidyBat) [inline mp/src/materialsystem/stdshaders/pbr_vs20.fxc]: Inconsistent space/tab usage in this file too.

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+    }
+
+    // initializing parameters
+    SHADER_INIT_PARAMS()
+    {
+        PBR_DX9_Vars_t info;
+        SetupVars(info);
+
+        // this is here for backwards compatibility
+        if (Q_strcmp(params[NORMALTEXTURE]->GetStringValue(), "<UNDEFINED>") != 0)
```
[the remark this answers is not in the document]
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: oh yeah I forgot about that, thanks

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+        {
+            params[BUMPMAP]->SetStringValue(params[NORMALTEXTURE]->GetStringValue());
+        }
+
+        // without this, dynamic lighting breaks
+        if (Q_strcmp(params[BUMPMAP]->GetStringValue(), "<UNDEFINED>") == 0)
```
[the remark this answers is not in the document]
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: gotcha

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
+        {
+            params[BUMPMAP]->SetStringValue("dev/flat_normal");
+        }
+
+        // check that the flashlight texture was set
+        Assert(info.FlashlightTexture >= 0);
```
[the remark this answers is not in the document]
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: I dunno, it's hacked together from other people's code, I'll remove it.

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+        {
+            params[FLASHLIGHTTEXTURE]->SetStringValue("effects/flashlight001");
+        }
+
+        // set material var2 flags
+        if (IS_FLAG_SET(MATERIAL_VAR_MODEL))
+        {
+            SET_FLAGS2(MATERIAL_VAR2_SUPPORTS_HW_SKINNING);             // required for skinning
+            SET_FLAGS2(MATERIAL_VAR2_LIGHTING_VERTEX_LIT);              // required for dynamic lighting
+            SET_FLAGS2(MATERIAL_VAR2_NEEDS_TANGENT_SPACES);             // required for dynamic lighting
```
[the remark this answers is not in the document]
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: fair enough, I'll do that

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp (hunk 2 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: good idea

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp (hunk 3 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: I thought I had already removed those, apparently not. I'll remove them now

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 4 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: I'll try

File: mp/src/materialsystem/stdshaders/pbr_ps30.fxc (hunk 5 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_ps30.fxc]: you're right, seems like an oversight

File: mp/src/materialsystem/stdshaders/pbr_vs20.fxc (hunk 6 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_vs20.fxc]: right, I'll try to fix it (again) (visual studio hates me)

File: mp/src/materialsystem/stdshaders/pbr_dx9.cpp (hunk 1 above)
Comment (bonjorno7) [reply mp/src/materialsystem/stdshaders/pbr_dx9.cpp]: sure, I'll do that

**Comment by bonjorno7** [comment]
I've adressed the changes pls

**Comment by McGravel** [comment]
> I've adressed the changes pls

But can you address your typo? 🤔

**Comment by bonjorno7** [comment]
Okay nvm for now

**Pull request closed without merging**